cli/src/commands/fix.rs

```diff
@@ -25,19 +28,37 @@ use jj_lib::repo_path::RepoPathBuf;
 use jj_lib::revset::{RevsetExpression, RevsetIteratorExt};
 use jj_lib::store::Store;
 use pollster::FutureExt;
+use rayon::iter::IntoParallelIterator;
+use rayon::prelude::ParallelIterator;
 use tracing::instrument;
 
 use crate::cli_util::{CommandHelper, RevisionArg};
-use crate::command_error::CommandError;
+use crate::command_error::{config_error_with_message, CommandError};
+use crate::config::CommandNameAndArgs;
 use crate::ui::Ui;
 
-/// Format files
+/// Update files with formatting fixes or other changes
+///
+/// The primary use case for this command is to apply the results of automatic
+/// code formatting tools to revisions that may not be properly formatted yet.
+/// It can also be used to modify files with other tools like `sed` or `sort`.
+///
+/// The changed files in the given revisions will be updated with any fixes
+/// determined by passing their file content through the external tool.
+/// Descendants will also be updated by passing their versions of the same files
+/// through the same external tool, which will never result in new conflicts.
+/// Files with existing conflicts will be updated on all sides of the conflict,
+/// which can potentially increase or decrease the number of conflict markers.
+///
+/// The external tool must accept the current file content on standard input,
+/// and return the updated file content on standard output. The output will not
+/// be used unless the tool exits with a successful exit code. Output on
+/// standard error will be ignored.
 #[derive(clap::Args, Clone, Debug)]
-#[command(verbatim_doc_comment)]
 pub(crate) struct FixArgs {
-    /// Fix these commits and all their descendants
+    /// Fix files in the specified revision(s) and their descendants
     #[arg(long, short)]
-    sources: Vec<RevisionArg>,
+    source: Vec<RevisionArg>,
```

Comment:
FWIW I prefer singular `--source` even for commands that accept multiple.

Tangential: git-branchless-test accepts an exact set of commits to fix, not a `--source` argument which includes their descendants, so I was surprised to see that interface here.

I spent a while trying to figure out why I was surprised. There is a meaningful interpretation of fixing a set of commits without its descendants: that you should rebase the descendant commits rather than reparent them.

I guess the main reason to do that would be for expensive or non-local fixes, to avoid repeating unnecessary work (such as lint fixes); or maybe to process commits that you know would fail to be fixed. It's not as useful for formatter-only fixes.

(There's no action required here, as I imagine we could easily add `-r` if we wanted it.)

Reply:
Minor thing I'd like to have is the no-op `jj fix -r` which just translates to `jj fix --source`.

Reply:
> Tangential: git-branchless-test accepts an exact set of commits to fix, not a `--source` argument which includes their descendants, so I was surprised to see that interface here.

The main reason for that is that it's easy to create conflicts by formatting a commit without its descendants. If a file is poorly formatted and you've been modifying that file many times in a chain, it's quite likely that `jj fix -r <early commit>` would result in conflicts later in the chain.

But maybe with jj's better understanding of conflict and/or the undo functionality, that's not much of a problem. Users can usually resolve those conflicts by simply formatting the commit (`jj fix -r` can even print a hint about that), or they can undo. So the reasons for `hg fix -r` that made us prefer `hg fix -s` don't really apply. I think I'm fine with dropping `--source` and adding `--revision` instead. What do you think, @arxanas? I'm also curious to hear what @hooper thinks.

Reply:
> Minor thing I'd like to have is the no-op `jj fix -r` which just translates to `jj fix --source`.

Can you elaborate? I haven't heard of that feature before.

Reply:
> Can you elaborate? I haven't heard of that feature before.

This is just the no-op `--revisions` flag from other commands which then get turned into sources. 

> (There's no action required here, as I imagine we could easily add `-r` if we wanted it.)

^, just the above but as an explicit request.

Reply:
> Minor thing I'd like to have is the no-op jj fix -r which just translates to jj fix --source.

I don't think I understand this; I'm expecting `-r` and `-s` to have different semantics, so how would you translate `-r` into `-s`?

---

> The main reason for that is that it's easy to create conflicts by formatting a commit without its descendants. If a file is poorly formatted and you've been modifying that file many times in a chain, it's quite likely that jj fix -r <early commit> would result in conflicts later in the chain.

To be clear, I don't think that `-s` is necessarily a problem/bad interface. I guess it depends on how often users would use `-r` vs `-s`, and whether they would get confused by the semantics/conflicts when you rebase descendants rather than reparent them.
- Would everyone just always be writing `jj fix -r 'set()::'`?
- Is `-r` a mistake for most users in most cases?
  - Probably yes for formatting, but maybe not for other kinds of fixes.
  - For formatting specifically, I think I rarely find the need to _not_ format certain commits. It's mostly for more-involved computations that it becomes useful.
- Is it appreciably different than how people would like to use `jj rebase`? There are definitely cases where you would want `-r`, but I think `-s` is more common.
  - I guess it's rather different in that `jj rebase` would fill in the resulting "holes"; I don't think we would expect `jj fix` to pull out the commits like that.

---

Actually, now that I consider it, I think git-branchless-test reparents all descendants rather than rebasing them (even when not included in the fix set), but I think _that's_ a bad interface and probably a footgun.

Reply:
It would be bad if we only offered `-r` and, when users didn't provide the full descendant set, it would _usually_ work, but only sometimes produce weird merge conflicts, depending on the specific contents of the files to be formatted. We should probably keep `-s` and use it by default in any documentation, similarly as we would do for `jj rebase`. Particularly when you consider that `jj fix` can't currently do non-formatting fixes, `-r` seems unnecessary for now.

Reply:
> It would be bad if we only offered `-r` and, when users didn't provide the full descendant set, it would _usually_ work, but only sometimes produce weird merge conflicts, depending on the specific contents of the files to be formatted.

Ack, I clearly didn't think that through. 

> Some users have reasons to want the alternative semantics of `-r`, which is why I intend to add that to `jj fix`.

SGTM.

> We have some prior art in proprietary code, but the basic idea is to fix the descendants in a more limited way instead of rebasing them. This means that the set of files and line ranges (line ranges being completely unimplemented in `jj fix` for now) would only consider what changed in `foo` and the formatted versions of `foo`, while ignoring changes in `foo..` when running `jj fix -r foo`.

This sounds like you have a [Weave] based implementation, which is needed for `jj absorb` anyway. Since you have plans to add it, can you also mention it on the issue?  

[Weave]: https://en.wikipedia.org/wiki/Interleaved_deltas

> The motivation for `-r foo` being different from `-r foo::` generally has to do with chains of commits that touch different parts of a repository, where the desired formatting is different in a way that isn't reflected by the current formatter configuration.

Yeah, dealing with external configuration layers is not something easily solvable at the source control layer, as it clearly requires a larger context. 

> So, I think my position is:
> 
> * `jj fix` is a worthwhile specialization of a subset of what `jj run` could do, because it is faster, more efficient, and simpler than materializing working copies and synchronizing subprocesses' access to working copies.
> * Better conflict handling is great, but it is orthogonal to this problem unless it includes language-aware auto-resolution (which, to be brief, would be difficult to accomplish).
> * `jj fix` should never rebase or "pull out" commits.
> * Most users want `-s` most of the time.
> * Some users need `-r` enough that not having it ruins the tool for them.
>   
>   * Introducing an `hg fix`-style `-r` would lock us in to that limitation.

I agree on all points.

> As for `--source` vs `--sources`, I think the plural has a slight benefit for the discoverability of it being a repeatable flag and evaluating each value to multiple revisions.

idc that much, as long as we keep it consistent across all commands.

Reply:
It sounds like everyone agrees that `-s` is good to have and probably the best default and best choice of initial implementation.

I also would very much prefer to keep `--source` to be consistent with other uses of `--source`. I don't think there's a reason that we shouldn't allow multiple sources for other commands, too?

(From a purely syntactic perspective, I don't think a pluralized option implies that the option can be repeated multiple times, anyways; I would expect it to accept an indefinite number of arguments without needing to repeat the option.)

I think having a specialized `fix` implementation is fine, but I don't think it makes sense to expose the same logical functionality to users in both `jj run` and `jj fix`. We should eventually add support to `jj fix` to operate on the working copy as well.

Reply:
Let's do `--source` for now for consistency.

Where is the best place to continue the conversation about the overall CLI of fix+run?

Reply:
Mentioned on discord that we will follow up in https://github.com/martinvonz/jj/issues/1869 and the fix-specific ones.